### PR TITLE
[SPARK-12508][PROJECT-INFRA] Fix minor bugs in dev/tests/pr_public_classes.sh script

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch fixes a handful of minor bugs in the `dev/tests/pr_public_classes.sh` script, which is used by the `run_tests_jenkins` script to detect the addition of new public classes:
- Account for differences between BSD and GNU `sed` in order to allow the script to run on OS X.
- Diff `$ghprbActualCommit^...$ghprbActualCommit` instead of `master...$ghprbActualCommit`: since `ghprbActualCommit` is a merge commit which results from merging the PR into the target branch, this will give us the desired diff and will avoid certain race-conditions which could lead to false-positives.
- Use `echo -e` instead of `echo` so that newline characters are handled correctly in output. This should fix a formatting glitch which caused the output to appear on a single line in the GitHub comment (see [the SC2028 page](https://github.com/koalaman/shellcheck/wiki/SC2028) on the Shellcheck wiki for more details).

### CODE REVIEW

**Comment by JoshRosen** [comment]
This should help to avoid this problem:

image

That comment (https://github.com/apache/spark/pull/10441#issuecomment-166806823) illustrates both a false-positive and the output formatting problem.

File: dev/tests/pr_public_classes.sh [hunk 1]
```diff
@@ -24,36 +24,44 @@
...
-  git diff master...$ghprbActualCommit --name-only  `# diff patch against master from branch point` \
+  git diff $ghprbActualCommit^...$ghprbActualCommit --name-only  `# diff patch against master from branch point` \
     | grep -v -e "\/test"                               `# ignore files in test directories` \
     | grep -e "\.py$" -e "\.java$" -e "\.scala$"        `# include only code files` \
     | tr "\n" " "
 )
+
 new_public_classes=$(
-  git diff master...$ghprbActualCommit ${source_files}      `# diff patch against master from branch point` \
+  git diff $ghprbActualCommit^...$ghprbActualCommit ${source_files}      `# diff patch against master from branch point` \
```
Comment (nchammas) [inline dev/tests/pr_public_classes.sh]: Huh... I didn't know you could do that.

I was aware of strange hiccups in the old logic but never figured it out. I guess this change makes the comment "diff against master from branch point" actually true now.

**Comment by nchammas** [comment]
LGTM.

cc @brennonyork since I believe he also worked on this script at one point when we were trying to replace it with some sbt functionality.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
(Re-testing to test a Jenkins conf change)

**Comment by JoshRosen** [comment]
I'm going to merge this into master.

**Pull request closed without merging**